src/repository.c

```diff
@@ -2477,6 +2477,65 @@ int git_repository_state_cleanup(git_repository *repo)
 	return git_repository__cleanup_files(repo, state_files, ARRAY_SIZE(state_files));
 }
 
+int git_repository__shallow_roots(git_array_oid_t *out, git_repository *repo)
+{
+	git_buf path = GIT_BUF_INIT;
+	git_buf contents = GIT_BUF_INIT;
+	int error, updated;
+	size_t start, end;
+
+	assert(out);
+
+	if ((error = git_buf_joinpath(&path, repo->path_repository, "shallow")) < 0)
+		return error;
+
+	error = git_futils_readbuffer_updated(&contents, git_buf_cstr(&path), &repo->shallow_checksum, &updated);
+	git_buf_free(&path);
+
+	if (error < 0)
+		return error;
+
+	if (!updated) {
+		goto unchanged;
+	}
+
+	git_array_clear(repo->shallow_oids);
+
+	start = end = 0;
+	while (end < contents.size) {
+		if (contents.ptr[end] == '\n' || end == contents.size) {
+			if (end - start == GIT_OID_HEXSZ) {
+				git_oid *oid = git_array_alloc(repo->shallow_oids);
+
+				error = git_oid_fromstrn(oid, contents.ptr + start, GIT_OID_HEXSZ);
+				if (error < 0)
+					return error;
+			}
+			start = end + 1;
+		}
+		end++;
```

Comment:
It seems like there's not a lot of error checking here...  It looks like it would handle a very malformed line if it was 80 characters of things that looked like hex characters...?  Also, walking every character yourself seems a bit unnecessary and I think that you could solve both problems with some minor refactoring.

I would rather see you use some existing standard library function (`git__strtok` may be a good choice) to find the terminating newline, validate that it's `GIT_OID_HEXSZ` away from your start, and then try to parse it...  

Does `git` handle blank linkes?  Comments?  Windows style newlines?  I assume not, but I wanted to ask.